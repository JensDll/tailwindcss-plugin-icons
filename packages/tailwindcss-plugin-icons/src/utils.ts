export function toKebabCase(str: string): string {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()
}

export function encodeSvg(svg: string) {
  return svg
    .replace(
      '<svg',
      ~svg.indexOf('xmlns') ? '<svg' : '<svg xmlns="http://www.w3.org/2000/svg"'
    )
    .replace(/"/g, "'")
    .replace(/%/g, '%25')
    .replace(/#/g, '%23')
    .replace(/{/g, '%7B')
    .replace(/}/g, '%7D')
    .replace(/</g, '%3C')
    .replace(/>/g, '%3E')
}

export interface IconifyJson {
  icons: {
    [iconName: string]: {
      body: string
      width?: number
      height?: number
    }
  }
  width: number
  height: number
}

export type IconMode = 'bg' | 'mask'

export function loadIcon(iconifyJson: IconifyJson, iconName: string) {
  let { width, height, icons } = iconifyJson
  let mode: IconMode | undefined

  iconName = iconName.replace(/\?(bg|mask)$/, (...values) => {
    mode = values[1]
    return ''
  })

  const icon = icons[iconName]
  const body = icon.body

  if (icon.width) {
    width = icon.width
  }
  if (icon.height) {
    height = icon.height
  }

  mode ??= body.includes('currentColor') ? 'mask' : 'bg'

  return {
    width,
    height,
    mode,
    body,
    normalizedIconName: iconName
  }
}
