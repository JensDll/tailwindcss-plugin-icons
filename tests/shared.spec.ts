import { html } from '~/scripts/utils.mjs'
import {
  isUri,
  loadIconFromIconifyJson,
  parseIconName,
  toKebabCase,
  encodeSvg,
  type ParsedIconName,
} from '~/src/chunks/shared/index.ts'

describe('toKebabCase', () => {
  test.each([
    {
      value: 'heroiconsSolid',
      expected: 'heroicons-solid',
    },
    {
      value: 'HeroiconsSolid',
      expected: 'heroicons-solid',
    },
    {
      value: 'heroicons-Solid',
      expected: 'heroicons-solid',
    },
  ])('$value', ({ value, expected }) => {
    const actual = toKebabCase(value)
    expect(actual).toBe(expected)
  })
})

describe('isUri', () => {
  test.each([
    {
      value: 'https://example.com',
      expected: true,
    },
    {
      value: 'http://example.com',
      expected: true,
    },
    {
      value: 'http:/example.com',
      expected: false,
    },
    {
      value: './src/icons.json',
      expected: false,
    },
    {
      value: 'src/icons.json',
      expected: false,
    },
    {
      value: '',
      expected: false,
    },
    {
      value: undefined,
      expected: false,
    },
    {
      value: null,
      expected: false,
    },
  ])('$value', ({ value, expected }) => {
    const actual = isUri(value)
    expect(actual).toBe(expected)
  })
})

describe('encodeSvg', () => {
  test('matches url encoded', () => {
    const svg = html`<svg
      viewBox="0 0 10 10"
      xmlns="http://www.w3.org/2000/svg"
    ></svg>`
    const actual = encodeSvg(svg)
    expect(actual).toMatchSnapshot()
  })

  test('without namespace', () => {
    const svg = html`<svg viewBox="0 0 10 10"></svg>`
    const actual = encodeSvg(svg)
    expect(actual).toMatchSnapshot()
  })

  test('without namespace and empty', () => {
    const svg = html`<svg></svg>`
    const actual = encodeSvg(svg)
    expect(actual).toMatchSnapshot()
  })

  test('needs xlink namespace', () => {
    const svg = html`<svg><use xlink:href="test"></use></svg>`
    const actual = encodeSvg(svg)
    expect(actual).toMatchSnapshot()
  })
})

describe('parseIconName', () => {
  test.each<{
    value: string
    expected: ParsedIconName
  }>([
    {
      value: 'plusCircle',
      expected: {
        normalizedIconName: 'plus-circle',
        iconMode: undefined,
      },
    },
    {
      value: 'plusCircle?bg',
      expected: {
        normalizedIconName: 'plus-circle',
        iconMode: 'bg',
      },
    },
    {
      value: 'plusCircle?mask',
      expected: {
        normalizedIconName: 'plus-circle',
        iconMode: 'mask',
      },
    },
  ])('$value', ({ value, expected }) => {
    const actual = parseIconName(value)
    expect(actual).toStrictEqual(expected)
  })
})

describe('loadIconFromJson', () => {
  const iconifyJson = {
    icons: {
      'current-color': {
        body: 'currentColor',
      },
      color: {
        body: '#fff',
      },
      left: {
        body: '',
        left: 10,
      },
      top: {
        body: '',
        top: 20,
      },
      width: {
        body: '',
        width: 30,
      },
      height: {
        body: '',
        height: 40,
      },
      'left+top': {
        body: '',
        left: 10,
        top: 20,
      },
      'left+top+width': {
        body: '',
        left: 10,
        top: 20,
        width: 30,
      },
      'left+top+width+height': {
        body: '',
        left: 10,
        top: 20,
        width: 30,
        height: 40,
      },
      rotate90: {
        body: 'body',
        rotate: 1,
        left: 5,
        top: 10,
        width: 50,
        height: 100,
      },
      rotate180: {
        body: 'body',
        rotate: 2,
        left: 5,
        top: 10,
        width: 50,
        height: 100,
      },
      'h-flip': {
        body: 'body',
        hFlip: true,
      },
      'v-flip': {
        body: 'body',
        vFlip: true,
      },
    },
    aliases: {
      'current-color-rotate90': {
        parent: 'current-color',
        rotate: 1,
      },
    },
    prefix: 'test',
    left: 1,
    top: 2,
    width: 3,
    height: 4,
  } as const

  type IconifyJson = typeof iconifyJson

  test.each<{
    iconName: `${keyof IconifyJson['icons'] | keyof IconifyJson['aliases']}${
      | '?bg'
      | '?mask'
      | ''}`
    expected: ReturnType<typeof loadIconFromIconifyJson>
  }>([
    {
      iconName: 'current-color',
      expected: {
        normalizedName: 'current-color',
        body: 'currentColor',
        mode: 'mask',
        left: 1,
        top: 2,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'current-color-rotate90',
      expected: {
        normalizedName: 'current-color-rotate90',
        body: '<g transform="rotate(90 2.5 4)">currentColor</g>',
        mode: 'mask',
        left: 1,
        top: 2,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'current-color?bg',
      expected: {
        normalizedName: 'current-color',
        body: 'currentColor',
        mode: 'bg',
        left: 1,
        top: 2,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'color',
      expected: {
        normalizedName: 'color',
        body: '#fff',
        mode: 'color',
        left: 1,
        top: 2,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'color?mask',
      expected: {
        normalizedName: 'color',
        body: '#fff',
        mode: 'mask',
        left: 1,
        top: 2,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'left',
      expected: {
        normalizedName: 'left',
        body: '',
        mode: 'color',
        left: 10,
        top: 2,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'top',
      expected: {
        normalizedName: 'top',
        body: '',
        mode: 'color',
        left: 1,
        top: 20,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'width',
      expected: {
        normalizedName: 'width',
        body: '',
        mode: 'color',
        left: 1,
        top: 2,
        width: 30,
        height: 4,
      },
    },
    {
      iconName: 'height',
      expected: {
        normalizedName: 'height',
        body: '',
        mode: 'color',
        left: 1,
        top: 2,
        width: 3,
        height: 40,
      },
    },
    {
      iconName: 'left+top',
      expected: {
        normalizedName: 'left+top',
        body: '',
        mode: 'color',
        left: 10,
        top: 20,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'left+top+width',
      expected: {
        normalizedName: 'left+top+width',
        body: '',
        mode: 'color',
        left: 10,
        top: 20,
        width: 30,
        height: 4,
      },
    },
    {
      iconName: 'left+top+width+height',
      expected: {
        normalizedName: 'left+top+width+height',
        body: '',
        mode: 'color',
        left: 10,
        top: 20,
        width: 30,
        height: 40,
      },
    },
    {
      iconName: 'rotate90',
      expected: {
        normalizedName: 'rotate90',
        body: '<g transform="rotate(90 30 60)">body</g>',
        mode: 'color',
        left: 5,
        top: 10,
        width: 50,
        height: 100,
      },
    },
    {
      iconName: 'rotate180',
      expected: {
        normalizedName: 'rotate180',
        body: '<g transform="rotate(180 30 60)">body</g>',
        mode: 'color',
        left: 5,
        top: 10,
        width: 50,
        height: 100,
      },
    },
    {
      iconName: 'h-flip',
      expected: {
        normalizedName: 'h-flip',
        body: '<g transform="translate(5 0) scale(-1 1)">body</g>',
        mode: 'color',
        left: 1,
        top: 2,
        width: 3,
        height: 4,
      },
    },
    {
      iconName: 'v-flip',
      expected: {
        normalizedName: 'v-flip',
        body: '<g transform="translate(0 8) scale(1 -1)">body</g>',
        mode: 'color',
        left: 1,
        top: 2,
        width: 3,
        height: 4,
      },
    },
  ])('$iconName', ({ iconName, expected }) => {
    const actual = loadIconFromIconifyJson(iconifyJson, iconName)
    expect(actual).toStrictEqual(expected)
  })
})
