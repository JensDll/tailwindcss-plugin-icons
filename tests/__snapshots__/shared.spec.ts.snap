// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encodeSvg > matches url encoded 1`] = `"%3Csvg viewBox='0 0 10 10' xmlns='http://www.w3.org/2000/svg' %3E%3C/svg%3E"`;

exports[`encodeSvg > needs xlink namespace 1`] = `"%3Csvg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3E%3Cuse xlink:href='test'%3E%3C/use%3E%3C/svg%3E"`;

exports[`encodeSvg > without namespace 1`] = `"%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 10 10'%3E%3C/svg%3E"`;

exports[`encodeSvg > without namespace and empty 1`] = `"%3Csvg xmlns='http://www.w3.org/2000/svg'%3E%3C/svg%3E"`;
